File: Java/refactoring1/src/main/java/za/co/entelect/refactoring1/domain/BankAccount.java
package za.co.entelect.refactoring1.domain;

public abstract class BankAccount {

    public static final double SAVINGS_CREDIT_INTEREST_RATE = 0.05D;
    public static final double SAVINGS_DEBIT_INTEREST_RATE = Double.NaN;

    public static final double MONEY_MARKET_CREDIT_INTEREST_RATE = 0.1D;
    public static final double MONEY_MARKET_DEBIT_INTEREST_RATE = Double.NaN;

    public static final double CHEQUE_CREDIT_INTEREST_RATE = 0.04D;
    public static final double CHEQUE_MARKET_DEBIT_INTEREST_RATE = 0.12D;

    private final double creditInterestsRate;
    private final double debitInterestRate;
    private Long balanceInCents;
    private boolean accountActive = true;

    public BankAccount(Long balanceInCents, double creditInterestsRate, double debitInterestRate) {
        this.balanceInCents = balanceInCents;
        this.creditInterestsRate = creditInterestsRate;
        this.debitInterestRate = debitInterestRate;
    }

    public abstract AccountType getAccountType();

    public long getBalanceInCents() {
        return balanceInCents;
    }

    public void updateBalance(long balanceInCents) {
        this.balanceInCents += balanceInCents;
    }

    public double getCreditInterestsRate() {
        return creditInterestsRate;
    }

    public double getDebitInterestRate() {
        return debitInterestRate;
    }

    public boolean isAccountActive() {
        return accountActive;
    }

    public void closeAccount() {
        this.accountActive = false;
    }

    public void reopenAccount() {
        this.accountActive = true;
    }
}

File: C#/Refactoring/refactoring_exercise_1/za/co/entelect/refactoring1/domain/BankAccount.cs
namespace refactoring_exercise_1.za.co.entelect.refactoring1.domain
{

    public abstract class BankAccount
    {

        public const double SAVINGS_CREDIT_INTEREST_RATE = 0.05D;
        public const double SAVINGS_DEBIT_INTEREST_RATE = double.NaN;
        public const long SAVINGS_ACCOUNT_FEE = 1000;

        public const double MONEY_MARKET_CREDIT_INTEREST_RATE = 0.1D;
        public const double MONEY_MARKET_DEBIT_INTEREST_RATE = double.NaN;
        public const long MONEY_MARKET_ACCOUNT_FEE = 1200;

        public const double CHEQUE_CREDIT_INTEREST_RATE = 0.04D;
        public const double CHEQUE_DEBIT_INTEREST_RATE = 0.12D;
        public const long CHEQUE_ACCOUNT_FEE = 1400;

        private readonly double creditInterestsRate;
        private readonly double debitInterestRate;
        private readonly long feeInCents;

        private long _balanceInCents;
        private bool _accountActive = true;

        public BankAccount(long balanceInCents, double creditInterestsRate, double debitInterestRate, long fee)
        {
            this._balanceInCents = balanceInCents;
            this.creditInterestsRate = creditInterestsRate;
            this.debitInterestRate = debitInterestRate;
            this.feeInCents = fee;
        }

        public abstract AccountType GetAccountType();

        public long BalanceInCents
        {
            get { return _balanceInCents; }
        }

        public void UpdateBalance(long balanceInCents)
        {
            this._balanceInCents += balanceInCents;
        }

        public double GetCreditInterestsRate()
        {
            return creditInterestsRate;
        }

        public double GetDebitInterestRate()
        {
            return debitInterestRate;
        }

        public bool IsAccountActive()
        {
            return _accountActive;
        }

        public void CloseAccount()
        {
            this._accountActive = false;
        }

        public void ReopenAccount()
        {
            this._accountActive = true;
        }

        public long GetFeeInCents()
        {
            return feeInCents;
        }
    }
}